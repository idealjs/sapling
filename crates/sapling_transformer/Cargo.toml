[package]
name = "sapling_transformer"
version = "0.1.0"
edition = "2024"
publish = false

[lib]
crate-type = ["cdylib", "rlib"]

[dependencies]
biome_js_syntax = { workspace = true }
biome_js_parser = { workspace = true }
biome_rowan.workspace = true
insta = { workspace = true }
camino = { workspace = true }
biome_js_formatter = { workspace = true }
biome_js_factory.workspace = true
biome_formatter = { workspace = true }
biome_js_semantic = { workspace = true }
wasm-bindgen = "0.2.100"
