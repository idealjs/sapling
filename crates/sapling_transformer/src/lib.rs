pub mod helpers;
pub mod scope;
pub mod transform_any_js_expression;
pub mod transform_any_js_object_member;
pub mod transform_any_js_statement;
pub mod transform_any_jsx_child;
pub mod transform_jsx_tag_expression_to_statements;
pub mod transformer;
pub mod transfrom_jsx_tag_expression;
pub mod write_transformation_snapshot;

pub use helpers::*;
pub use transform_any_jsx_child::*;
pub use transform_jsx_tag_expression_to_statements::*;

use crate::transformer::{Config, SaplingTransformer};
use crate::transfrom_jsx_tag_expression::TransformAnyJsxTagExpressionOptions;
use biome_formatter::IndentStyle;
use biome_js_formatter::context::JsFormatOptions;
use biome_js_formatter::format_node;
use biome_js_parser::{JsParserOptions, parse};
use biome_js_semantic::{BindingExtensions, SemanticModelOptions, semantic_model};
use biome_js_syntax::JsFileSource;
use biome_js_syntax::{AnyJsDecorator, JsClassDeclaration, JsSyntaxKind, JsxTagExpression};
use biome_rowan::AstNode;
use biome_rowan::SyntaxNodeCast;
use biome_rowan::SyntaxRewriter;
use std::collections::{HashMap, HashSet};
use wasm_bindgen::prelude::*;

impl biome_rowan::SyntaxRewriter for SaplingTransformer<'_> {
    type Language = biome_js_syntax::JsLanguage;

    fn visit_node(
        &mut self,
        node: biome_rowan::SyntaxNode<Self::Language>,
    ) -> biome_rowan::VisitNodeSignal<Self::Language> {
        use biome_rowan::VisitNodeSignal;

        match node.kind() {
            JsSyntaxKind::JS_CLASS_DECLARATION => {
                if let Some(class_node) = node.clone().cast::<JsClassDeclaration>() {
                    let _ = class_node
                        .members()
                        .into_iter()
                        .try_for_each(|member| -> Option<()> {
                            let member = match member.as_js_property_class_member() {
                                Some(v) => v,
                                None => return Some(()),
                            };

                            let is_reactive = member.modifiers().into_iter().any(|modifier| {
                                let modifier = match modifier.as_js_decorator() {
                                    Some(v) => v,
                                    None => return false,
                                };
                                let expr = match modifier.expression().ok() {
                                    Some(v) => v,
                                    None => return false,
                                };

                                let binding = match expr {
                                    AnyJsDecorator::JsIdentifierExpression(expr) => {
                                        match expr.name().ok() {
                                            Some(name_node) => {
                                                name_node.binding(&self.semantic_model)
                                            }
                                            None => return false,
                                        }
                                    }
                                    AnyJsDecorator::JsStaticMemberExpression(expr) => {
                                        match expr.object().ok() {
                                            Some(obj) => {
                                                if let Some(ident_expr) =
                                                    obj.as_js_identifier_expression()
                                                {
                                                    match ident_expr.name().ok() {
                                                        Some(name_node) => {
                                                            name_node.binding(&self.semantic_model)
                                                        }
                                                        None => return false,
                                                    }
                                                } else {
                                                    None
                                                }
                                            }
                                            None => return false,
                                        }
                                    }
                                    _ => None,
                                };

                                let binding = if let Some(binding) = binding {
                                    binding
                                } else {
                                    return false;
                                };

                                if get_js_module_source_from_binding(&self.semantic_model, &binding)
                                    .as_deref()
                                    == Some("@idealjs/sapling")
                                {
                                    return true;
                                }
                                return false;
                            });

                            let member_key = member
                                .name()
                                .ok()?
                                .as_js_literal_member_name()?
                                .name()
                                .ok()?
                                .text()
                                .to_string();
                            if is_reactive {
                                self.decorated_members.insert(member_key);
                            }
                            Some(())
                        });
                }
                VisitNodeSignal::Traverse(node)
            }
            JsSyntaxKind::JSX_TAG_EXPRESSION => {
                if let Some(jsx_node) = node.clone().cast::<JsxTagExpression>() {
                    // build transformer for this replacement

                    if let Some(new_expr) = self.transform_jsx_tag_expression(
                        &jsx_node,
                        TransformAnyJsxTagExpressionOptions { parent_id: None },
                    ) {
                        let new_syntax = new_expr.syntax().clone();
                        return VisitNodeSignal::Traverse(new_syntax);
                    }
                }
                VisitNodeSignal::Traverse(node)
            }
            _ => VisitNodeSignal::Traverse(node),
        }
    }
}

#[wasm_bindgen]
pub fn transform(input_code: String) -> Option<String> {
    // parse
    let parsed_root = parse(
        input_code.as_str(),
        JsFileSource::tsx(),
        JsParserOptions::default(),
    );
    let js_tree = parsed_root.try_tree()?;

    let semantic_model = semantic_model(&js_tree, SemanticModelOptions::default());

    // First pass: collect decorated members per class and build the bit map.
    // We keep a map from class text range -> set of decorated member names.
    let mut decorated_members: HashSet<String> = HashSet::new();
    let mut string_tree = StringTree::default();

    // run rewriter
    let root = js_tree.into_syntax();
    let mut transformer = SaplingTransformer {
        semantic_model: semantic_model,
        scope_generated_identifiers: HashMap::new(),
        config: Config::default(),
        decorated_members: &mut decorated_members,
        string_tree: &mut string_tree,
    };

    let syntax_node = transformer.transform(root);

    // format and return
    let formatted = format_node(
        JsFormatOptions::new(JsFileSource::default()).with_indent_style(IndentStyle::Space),
        &syntax_node,
    )
    .ok()?
    .print()
    .ok()?
    .as_code()
    .to_string();
    Some(formatted)
}
