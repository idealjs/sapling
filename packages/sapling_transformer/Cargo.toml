[package]
name = "sapling_transformer"
version = "0.1.0"
edition = "2024"
publish = false

[dependencies]
sapling_macros = { path = "../sapling_macros" }
sapling_shared = { path = "../sapling_shared" }
biome_js_syntax = { workspace = true }
biome_js_parser = { workspace = true }
biome_rowan.workspace = true
insta = { workspace = true }
biome_analyze = { workspace = true }
biome_js_transform = { workspace = true }
biome_diagnostics = { workspace = true }
camino = { workspace = true }
biome_js_formatter = { workspace = true }
biome_test_utils = { workspace = true }
tests_macros = { workspace = true }
biome_js_factory.workspace = true
biome_formatter = { workspace = true }
biome_js_semantic = { workspace = true }
