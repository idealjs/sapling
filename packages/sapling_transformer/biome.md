# Biome CST 遍历和转换

## CST 遍历机制

Biome 通过访问者模式实现 CST 的遍历，核心包括：

1. WalkEvent 遍历事件:
   - Enter(T) - 进入节点时触发
   - Leave(T) - 离开节点后触发

2. 遍历器类型:
   - Preorder - 只遍历节点
   - PreorderWithTokens - 遍历节点和 Token
   - PreorderTokens - 只遍历 Token

3. 访问者模式：
   - Visitor trait - 全局访问者，可访问任意节点
   - NodeVisitor trait - 节点特定访问者，生命周期与节点访问绑定
   - merge_node_visitors! 宏用于组合多个 NodeVisitor

## 代码转换实现

### 基本转换结构

1. 定义转换规则:
```rust
declare_transformation! {
    pub(crate) MyTransform {
        version: "1.0.0",
        name: "transformExample",
        language: "js",
    }
}
```

2. 实现 Rule trait:
```rust
impl Rule for MyTransform {
    type Query = Ast<TargetNode>;
    type State = TransformState;

    fn run(ctx: &RuleContext<Self>) -> Self::Signals {
        // 收集转换所需信息
    }

    fn transform(ctx: &RuleContext<Self>, state: &Self::State) -> ption<JsBatchMutation> {
        // 执行转换
    }
}
```

### 实际转换案例

1. TypeScript enum 转换
```typescript
// 输入
enum Foo {
    A,
    B
}

// 输出
var Foo;
(function (Foo) {
    Foo[Foo["A"] = 0] = "A";
    Foo[Foo["B"] = 1] = "B";
})(Foo || (Foo = {}));
```

2. Solid.js JSX 转换
```jsx
// 输入
function Counter() {
  const [count, setCount] = createSignal(1);
  return (
    <div>
      <For each={items}>
        {(item) => (
          <div>
            <span>{item.name}</span>
            <span>{item.value}</span>
          </div>
        )}
      </For>
    </div>
  );
}

// 输出 (客户端渲染)
var _el$ = _tmpl$();
_$insert(_el$, _$createComponent(For, {
  each: items,
  children: item => (() => {
    var _el$2 = _tmpl$2(),
      _el$3 = _el$2.firstChild,
      _el$4 = _el$3.nextSibling;
    _$insert(_el$3, () => item.name);
    _$insert(_el$4, () => item.value);
    return _el$2;
  })()
}));
return _el$;

// 输出 (服务端渲染)
import { ssr as _$ssr } from "solid-js/web";
import { escape as _$escape } from "solid-js/web";
var _tmpl$ = ["<div", ">", "</div>"],
    _tmpl$2 = ["<div", "><span>", "</span><span>", "</span></div>"];

function Counter() {
  const [count, setCount] = createSignal(1);
  return _$ssr(_tmpl$, _$ssrHydrationKey(), 
    _$escape(_$createComponent(For, {
      each: items,
      children: item => _$ssr(_tmpl$2, _$ssrHydrationKey(), 
        _$escape(item.name), 
        _$escape(item.value)
      )
    }))
  );
}
```

### 转换实现要点

1. 状态管理:
   - 追踪模板和变量
   - 维护作用域信息
   - 管理导入声明

2. 节点转换:
   - 使用 factory 函数创建新节点
   - 通过 mutation API 替换节点
   - 处理子节点递归转换

3. 关键考虑:
   - 类型安全
   - 作用域处理
   - 正确的错误处理
   - 源码映射支持

这种基于访问者模式的转换机制可以实现从简单的语法转换到复杂的框架编译，同时保证类型安全和可维护性。

# Biome 相关内容

### 转换器实现

1. 收集模板和状态:
```rust
struct SolidTransformState {
    // 追踪模板变量
    template_vars: Vec<String>,
    // 追踪作用域内的变量
    scope_vars: Vec<String>,
    // 收集的模板定义
    templates: HashMap<String, JsxElement>,
}
```

2. 实现转换:
```rust 
impl Rule for SolidTransform {
    type Query = Ast<JsxElement>;
    type State = SolidTransformState;

    fn run(ctx: &RuleContext<Self>) -> Self::Signals {
        let node = ctx.query();
        // 收集静态模板和状态
        Some(collect_transform_info(node))
    }

    fn transform(ctx: &RuleContext<Self>, state: &Self::State) -> Option<JsBatchMutation> {
        let node = ctx.query();
        let mut mutation = node.clone().begin();

        match state.mode {
            // 客户端渲染转换
            ClientSide => transform_client_side(node, state, &mut mutation),
            // 服务端渲染转换 
            ServerSide => transform_server_side(node, state, &mut mutation),
        }

        Some(mutation)
    }
}
```

3. 辅助函数:
```rust
// 生成模板引用
fn make_template_ref(name: &str) -> JsCallExpression {
    js_call_expression(
        js_identifier("_tmpl$"),
        []
    )
}

// 生成插入调用
fn make_insert_call(target: &str, value: JsExpression) -> JsCallExpression {
    js_call_expression(
        js_identifier("_$insert"),
        [js_identifier(target), value]
    )
}

// SSR 相关转换
fn make_ssr_call(template: &str, key: JsExpression, children: JsExpression) -> JsCallExpression {
    js_call_expression(
        js_identifier("_$ssr"),
        [
            js_identifier(template),
            make_hydration_key_call(),
            make_escape_call(children)
        ]
    )
}
```

这种转换设计可以:
1. 处理复杂的 JSX 结构
2. 支持客户端和服务端两种模式
3. 正确处理作用域和模板
4. 维护代码的可读性和可调试性

