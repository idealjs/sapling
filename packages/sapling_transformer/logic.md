```mermaid
flowchart TD
    A[输入: JsModule, TransformState] --> B[遍历]
    B --> C{transform_module_item_with_tracker}
    C --> D[判断节点类型]
    D -- JSX/表达式/函数/变量等 --> E[transform_statement_with_tracker/transform_expression_with_tracker/transform_arrow_function]
    E -- JSX 节点 --> F[create_solidjs_call_with_tracker/with_tracker_self_closing]
    F -- 子节点递归 --> E
    E -- 其它复合节点递归 --> E1{分发}
    E1 -- 表达式语句/return语句 --> E2[transform_expression_with_tracker]
    E1 -- 函数声明 --> E3[transform_statement_with_tracker]
    E1 -- 变量声明(箭头函数) --> E4[transform_arrow_function]
    E1 -- 变量声明(JSX/表达式) --> E2
    E2 -- JSX/括号表达式/嵌套箭头函数 --> E2/E4
    E3 -- 递归处理函数体语句 --> E3
    E4 -- 表达式体/return语句 --> E2
    E -- 基础节点 --> G[统计 HelperUsageTracker]
    G --> H[返回转换后模块项]
    H -- 循环遍历所有 module.items() --> C
    H --> I[收集所有转换后模块项]
    I --> J[generate_solid_imports]
    J --> K[插入 import 语句]
    K --> L[重建 JsModule AST]
    L --> M[输出: 新 JsModule]

```
