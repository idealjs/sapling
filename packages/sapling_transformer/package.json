{
  "name": "sapling_transformer",
  "packageManager": "yarn@4.0.1"
}
