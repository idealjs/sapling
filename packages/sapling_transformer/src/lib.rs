use biome_js_syntax::JsLanguage;
use biome_rowan::BatchMutation;

pub mod compatible;
pub mod create_statement;
pub mod scope;
pub mod transformer;
pub mod write_transformation_snapshot;

pub use compatible::*;
pub use create_statement::*;
pub use scope::*;
pub use transformer::*;
pub use write_transformation_snapshot::*;

pub type JsBatchMutation = BatchMutation<JsLanguage>;
