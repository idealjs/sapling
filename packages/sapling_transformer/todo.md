# SaplingJS Transformer Development TODO

## 当前状态 ✅ (更新)

### 已完成功能
- [x] JSX 元素解析与识别
- [x] 基础元素转换为 `_$createElement()` 调用
- [x] 文本节点处理 (`_$insertNode` + `_$createTextNode`)
- [x] 简单表达式子元素处理
- [x] IIFE (立即执行函数) 包装
- [x] 编译错误修复完成
- [x] **SolidJS Transform 框架架构** - 新增！
  - [x] 从 `JsModule` 根节点开始的转换逻辑
  - [x] `TransformState` 状态管理
  - [x] 模块级别的 JSX 检测和收集框架
  - [x] 整个模块替换的批量变更机制

## SolidJS Transform 架构

```mermaid
flowchart TD
    A[JsModule 输入] --> B[Rule.run: 检测 JSX]
    B --> C{包含 JSX?}
    C -->|是| D[收集 JSX 元素信息]
    C -->|否| E[跳过转换]
    D --> F[创建 TransformState]
    F --> G[Rule.transform: 转换模块]
    G --> H[生成导入语句]
    G --> I[转换 JSX 表达式]
    G --> J[重建 JsModule]
    J --> K[replace_node 整个模块]
```

### 核心解决方案：根节点替换
对于你提出的"JsModule 是根节点，如何 replace"的问题，我们采用了以下方案：

```rust
// 在 transform 方法中
let mut mutation = module.clone().begin();
mutation.replace_node(module.clone(), new_module);
```

这样可以直接替换整个模块，因为 `BatchMutation` 机制允许根节点的替换。

---

## 下一步实现优先级

### 1. 高优先级 - 完善框架核心逻辑

#### 1.1 实现 JSX 检测逻辑
- [ ] 完善 `contains_jsx()` 函数
  - [ ] 递归遍历 AST 查找 `JsxTagExpression` 
  - [ ] 支持函数内、表达式内的 JSX 检测

#### 1.2 实现 JSX 收集逻辑
- [ ] 完善 `collect_jsx_elements()` 函数
  - [ ] 收集 JSX 元素的位置信息
  - [ ] 记录 JSX 的上下文信息（函数内、return 语句等）

#### 1.3 实现模块转换逻辑
- [ ] 完善 `transform_module()` 函数
  - [ ] 生成必要的导入语句
  - [ ] 转换包含 JSX 的语句/表达式
  - [ ] 重新构建完整的 `JsModule`

### 2. 中优先级 - 属性和高级特性

#### 2.1 属性处理实现
- [ ] 完善 `handle_jsx_attributes()` 函数
- [ ] 支持静态属性、动态属性、事件处理

#### 2.2 导入语句生成
- [ ] 完善 `generate_solid_imports()` 函数
- [ ] 自动检测需要的运行时函数并生成对应导入

### 3. 低优先级 - 测试和优化

#### 3.1 测试覆盖
- [ ] 添加模块级别的转换测试
- [ ] 验证根节点替换的正确性

---

## 技术要点记录

### 关键设计决策
1. **Query 类型**: 从 `Ast<JsxTagExpression>` 改为 `Ast<JsModule>`，实现整个模块的转换
2. **State 管理**: 使用 `TransformState` 收集和传递转换所需的信息
3. **根节点替换**: 通过 `BatchMutation` 直接替换整个模块，解决根节点无父节点的问题

### 当前框架状态
- ✅ 编译通过，无语法错误
- ✅ 类型系统兼容
- ⚠️ 核心逻辑函数需要实现（标记为 TODO）

### 推荐下一步
优先实现 `contains_jsx()` 函数，这是整个转换流程的入口点。可以先实现一个简单版本来验证框架的可行性。