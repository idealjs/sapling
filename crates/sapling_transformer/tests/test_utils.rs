use camino::Utf8Path;
use sapling_transformer::{
    transform, write_transformation_snapshot::write_transformation_snapshot,
};
use std::fs::read_to_string;

pub fn run_test(input: &'static str) -> Option<()> {
    let input_file = Utf8Path::new(input);
    let file_name = input_file.file_name().unwrap();

    let mut snapshot = String::new();

    let input_code = read_to_string(input_file)
        .unwrap_or_else(|err| panic!("failed to read {input_file:?}: {err:?}"));

    let formatted = transform(input_code.clone())?;

    write_transformation_snapshot(
        &mut snapshot,
        input_code.clone().as_str(),
        formatted.as_str(),
        input_file.extension()?,
    );

    insta::with_settings!({
        prepend_module_to_snapshot => false,
        snapshot_path => {
            let path = input_file.parent().unwrap();
            let path_str = path.as_str();
            if let Some(idx) = path_str.find("specs/") {
                Utf8Path::new(&path_str[idx..])
            } else {
                path
            }
        },
    },
    {
        insta::assert_snapshot!(file_name, snapshot, file_name);
    });

    None
}
