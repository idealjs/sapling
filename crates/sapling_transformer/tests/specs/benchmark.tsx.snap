---
source: crates/sapling_transformer/tests/test_utils.rs
expression: benchmark.tsx
---
# Input
```tsx
import { Mutate, State } from "@idealjs/sapling";

import For from "./for";

const A = [
  "pretty",
  "large",
  "big",
  "small",
  "tall",
  "short",
  "long",
  "handsome",
  "plain",
  "quaint",
  "clean",
  "elegant",
  "easy",
  "angry",
  "crazy",
  "helpful",
  "mushy",
  "odd",
  "unsightly",
  "adorable",
  "important",
  "inexpensive",
  "cheap",
  "expensive",
  "fancy",
];

const C = [
  "red",
  "yellow",
  "blue",
  "green",
  "pink",
  "brown",
  "purple",
  "brown",
  "white",
  "black",
  "orange",
];

const N = [
  "table",
  "chair",
  "house",
  "bbq",
  "desk",
  "car",
  "pony",
  "cookie",
  "sandwich",
  "burger",
  "pizza",
  "mouse",
  "keyboard",
];

const random = (max: number) => Math.round(Math.random() * 1000) % max;

let nextId = 1;

function assignData(data: { id: number; label: string }[], offset = 0) {
  const count = data.length;
  for (let i = offset; i < count; i++) {
    data[i] = {
      id: nextId++,
      label: `${A[random(A.length)]} ${C[random(C.length)]} ${N[random(N.length)]}`,
    };
  }
  return data;
}

interface IData {
  id: number;
  label: string;
}

class Row {
  @State accessor props: {
    item: IData;
    selected: number;
    onSelect: (id: number) => void;
    onRemove: (id: number) => void;
  };
  constructor(p: {
    item: IData;
    selected: number;
    onSelect: (id: number) => void;
    onRemove: (id: number) => void;
  }) {
    this.props = p;
  }
  render() {
    const { item, selected, onSelect, onRemove } = this.props;
    const isSelected = selected === item.id;
    return (
      <tr className={isSelected ? "danger" : ""}>
        <td className="col-md-1">{item.id}</td>
        <td className="col-md-4">
          <a onClick={() => onSelect(item.id)}>{item.label}</a>
        </td>
        <td className="col-md-1">
          <a onClick={() => onRemove(item.id)}>
            <span
              className="glyphicon glyphicon-remove"
              aria-hidden="true"
            ></span>
          </a>
        </td>
        <td className="col-md-6"></td>
      </tr>
    );
  }
}

class App {
  @State accessor data: IData[] = [];
  @State accessor selected: number = 0;

  @Mutate
  run() {
    this.data = new Array(1000);
    assignData(this.data);
    this.selected = 0;
  }

  @Mutate
  runLots() {
    this.data = new Array(10000);
    assignData(this.data);
    this.selected = 0;
  }

  @Mutate
  add() {
    const offset = this.data.length;
    this.data.length += 1000;
    assignData(this.data, offset);
    this.selected = 0;
  }

  @Mutate
  update() {
    for (let i = 0; i < this.data.length; i += 10) {
      const r = this.data[i];
      this.data[i] = { id: r.id, label: r.label + " !!!" };
    }
  }

  @Mutate
  remove(id: number) {
    const idx = this.data.findIndex((d) => d.id === id);
    if (idx >= 0) this.data.splice(idx, 1);
  }

  @Mutate
  select(id: number) {
    this.selected = id;
  }

  @Mutate
  clear() {
    this.data = [];
    this.selected = 0;
  }

  @Mutate
  swapRows() {
    if (this.data.length > 998) {
      const tmp = this.data[1];
      this.data[1] = this.data[998];
      this.data[998] = tmp;
    }
  }

  render() {
    const { data, selected } = this;
    return (
      <div className="container">
        <div className="jumbotron">
          <div className="row">
            <div className="col-md-6">
              <h1>Sapling</h1>
            </div>
            <div className="col-md-6">
              <div className="row">
                <div className="col-sm-6 smallpad">
                  <button
                    type="button"
                    className="btn btn-primary btn-block"
                    id="run"
                    onClick={() => this.run()}
                  >
                    Create 1,000 rows
                  </button>
                </div>
                <div className="col-sm-6 smallpad">
                  <button
                    type="button"
                    className="btn btn-primary btn-block"
                    id="runlots"
                    onClick={() => this.runLots()}
                  >
                    Create 10,000 rows
                  </button>
                </div>
                <div className="col-sm-6 smallpad">
                  <button
                    type="button"
                    className="btn btn-primary btn-block"
                    id="add"
                    onClick={() => this.add()}
                  >
                    Append 1,000 rows
                  </button>
                </div>
                <div className="col-sm-6 smallpad">
                  <button
                    type="button"
                    className="btn btn-primary btn-block"
                    id="update"
                    onClick={() => this.update()}
                  >
                    Update every 10th row
                  </button>
                </div>
                <div className="col-sm-6 smallpad">
                  <button
                    type="button"
                    className="btn btn-primary btn-block"
                    id="clear"
                    onClick={() => this.clear()}
                  >
                    Clear
                  </button>
                </div>
                <div className="col-sm-6 smallpad">
                  <button
                    type="button"
                    className="btn btn-primary btn-block"
                    id="swaprows"
                    onClick={() => this.swapRows()}
                  >
                    Swap Rows
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>

        <table className="table table-hover table-striped test-data">
          <tbody>
            <For each={data} fallback={<tr />}>
              {(item) => (
                <Row
                  item={item}
                  selected={selected}
                  onSelect={(id) => this.select(id)}
                  onRemove={(id) => this.remove(id)}
                />
              )}
            </For>
          </tbody>
        </table>
        <span
          className="preloadicon glyphicon glyphicon-remove"
          aria-hidden="true"
        ></span>
      </div>
    );
  }
}

export default App;

```

# Transformations
```tsx
import { Mutate, State } from "@idealjs/sapling";

import For from "./for";

const A = [
  "pretty",
  "large",
  "big",
  "small",
  "tall",
  "short",
  "long",
  "handsome",
  "plain",
  "quaint",
  "clean",
  "elegant",
  "easy",
  "angry",
  "crazy",
  "helpful",
  "mushy",
  "odd",
  "unsightly",
  "adorable",
  "important",
  "inexpensive",
  "cheap",
  "expensive",
  "fancy",
];

const C = [
  "red",
  "yellow",
  "blue",
  "green",
  "pink",
  "brown",
  "purple",
  "brown",
  "white",
  "black",
  "orange",
];

const N = [
  "table",
  "chair",
  "house",
  "bbq",
  "desk",
  "car",
  "pony",
  "cookie",
  "sandwich",
  "burger",
  "pizza",
  "mouse",
  "keyboard",
];

const random = (max: number) => Math.round(Math.random() * 1000) % max;

let nextId = 1;

function assignData(data: { id: number; label: string }[], offset = 0) {
  const count = data.length;
  for (let i = offset; i < count; i++) {
    data[i] = {
      id: nextId++,
      label: `${A[random(A.length)]} ${C[random(C.length)]} ${N[random(N.length)]}`,
    };
  }
  return data;
}

interface IData {
  id: number;
  label: string;
}

class Row {
  @State accessor props: {
    item: IData;
    selected: number;
    onSelect: (id: number) => void;
    onRemove: (id: number) => void;
  };
  constructor(p: {
    item: IData;
    selected: number;
    onSelect: (id: number) => void;
    onRemove: (id: number) => void;
  }) {
    this.props = p;
  }
  render() {
    const { item, selected, onSelect, onRemove } = this.props;
    const isSelected = selected === item.id;
    return _$createJsxTagElement(() => {
      let _el$ = _$createElement("tr");
      _$effect(
        () => {
          _$setProp(_el$, "className", isSelected ? "danger" : "");
        },
        () => {
          return [isSelected ? "danger" : ""];
        },
      );
      let _el$1 = _$createElement("td");
      _$setProp(_el$1, "className", "col-md-1");
      _$effect(
        () => {
          _$insert(_el$1, item.id);
        },
        () => {
          return [item.id];
        },
      );
      _$insertNode(_el$, _el$1);
      let _el$2 = _$createElement("td");
      _$setProp(_el$2, "className", "col-md-4");
      let _el$3 = _$createElement("a");
      _$setProp(_el$3, "onClick", () => onSelect(item.id));
      _$effect(
        () => {
          _$insert(_el$3, item.label);
        },
        () => {
          return [item.label];
        },
      );
      _$insertNode(_el$2, _el$3);
      _$insertNode(_el$, _el$2);
      let _el$4 = _$createElement("td");
      _$setProp(_el$4, "className", "col-md-1");
      let _el$5 = _$createElement("a");
      _$setProp(_el$5, "onClick", () => onRemove(item.id));
      let _el$6 = _$createElement("span");
      _$setProp(_el$6, "className", "glyphicon glyphicon-remove");
      _$setProp(_el$6, "aria-hidden", "true");
      _$insertNode(_el$5, _el$6);
      _$insertNode(_el$4, _el$5);
      _$insertNode(_el$, _el$4);
      let _el$7 = _$createElement("td");
      _$setProp(_el$7, "className", "col-md-6");
      _$insertNode(_el$, _el$7);
      return _el$;
    });
  }
}

class App {
  @State accessor data: IData[] = [];
  @State accessor selected: number = 0;

  @Mutate
  run() {
    this.data = new Array(1000);
    assignData(this.data);
    this.selected = 0;
  }

  @Mutate
  runLots() {
    this.data = new Array(10000);
    assignData(this.data);
    this.selected = 0;
  }

  @Mutate
  add() {
    const offset = this.data.length;
    this.data.length += 1000;
    assignData(this.data, offset);
    this.selected = 0;
  }

  @Mutate
  update() {
    for (let i = 0; i < this.data.length; i += 10) {
      const r = this.data[i];
      this.data[i] = { id: r.id, label: r.label + " !!!" };
    }
  }

  @Mutate
  remove(id: number) {
    const idx = this.data.findIndex((d) => d.id === id);
    if (idx >= 0) this.data.splice(idx, 1);
  }

  @Mutate
  select(id: number) {
    this.selected = id;
  }

  @Mutate
  clear() {
    this.data = [];
    this.selected = 0;
  }

  @Mutate
  swapRows() {
    if (this.data.length > 998) {
      const tmp = this.data[1];
      this.data[1] = this.data[998];
      this.data[998] = tmp;
    }
  }

  render() {
    const { data, selected } = this;
    return _$createJsxTagElement(() => {
      let _el$ = _$createElement("div");
      _$setProp(_el$, "className", "container");
      let _el$1 = _$createElement("div");
      _$setProp(_el$1, "className", "jumbotron");
      let _el$2 = _$createElement("div");
      _$setProp(_el$2, "className", "row");
      let _el$3 = _$createElement("div");
      _$setProp(_el$3, "className", "col-md-6");
      let _el$4 = _$createElement("h1");
      _$insertNode(_el$4, _$createTextNode("Sapling"));
      _$insertNode(_el$3, _el$4);
      _$insertNode(_el$2, _el$3);
      let _el$5 = _$createElement("div");
      _$setProp(_el$5, "className", "col-md-6");
      let _el$6 = _$createElement("div");
      _$setProp(_el$6, "className", "row");
      let _el$7 = _$createElement("div");
      _$setProp(_el$7, "className", "col-sm-6 smallpad");
      let _el$8 = _$createElement("button");
      _$setProp(_el$8, "type", "button");
      _$setProp(_el$8, "className", "btn btn-primary btn-block");
      _$setProp(_el$8, "id", "run");
      _$setProp(_el$8, "onClick", () => this.run());
      _$insertNode(_el$8, _$createTextNode("Create 1,000 rows"));
      _$insertNode(_el$7, _el$8);
      _$insertNode(_el$6, _el$7);
      let _el$9 = _$createElement("div");
      _$setProp(_el$9, "className", "col-sm-6 smallpad");
      let _el$10 = _$createElement("button");
      _$setProp(_el$10, "type", "button");
      _$setProp(_el$10, "className", "btn btn-primary btn-block");
      _$setProp(_el$10, "id", "runlots");
      _$setProp(_el$10, "onClick", () => this.runLots());
      _$insertNode(_el$10, _$createTextNode("Create 10,000 rows"));
      _$insertNode(_el$9, _el$10);
      _$insertNode(_el$6, _el$9);
      let _el$11 = _$createElement("div");
      _$setProp(_el$11, "className", "col-sm-6 smallpad");
      let _el$12 = _$createElement("button");
      _$setProp(_el$12, "type", "button");
      _$setProp(_el$12, "className", "btn btn-primary btn-block");
      _$setProp(_el$12, "id", "add");
      _$setProp(_el$12, "onClick", () => this.add());
      _$insertNode(_el$12, _$createTextNode("Append 1,000 rows"));
      _$insertNode(_el$11, _el$12);
      _$insertNode(_el$6, _el$11);
      let _el$13 = _$createElement("div");
      _$setProp(_el$13, "className", "col-sm-6 smallpad");
      let _el$14 = _$createElement("button");
      _$setProp(_el$14, "type", "button");
      _$setProp(_el$14, "className", "btn btn-primary btn-block");
      _$setProp(_el$14, "id", "update");
      _$setProp(_el$14, "onClick", () => this.update());
      _$insertNode(_el$14, _$createTextNode("Update every 10th row"));
      _$insertNode(_el$13, _el$14);
      _$insertNode(_el$6, _el$13);
      let _el$15 = _$createElement("div");
      _$setProp(_el$15, "className", "col-sm-6 smallpad");
      let _el$16 = _$createElement("button");
      _$setProp(_el$16, "type", "button");
      _$setProp(_el$16, "className", "btn btn-primary btn-block");
      _$setProp(_el$16, "id", "clear");
      _$setProp(_el$16, "onClick", () => this.clear());
      _$insertNode(_el$16, _$createTextNode("Clear"));
      _$insertNode(_el$15, _el$16);
      _$insertNode(_el$6, _el$15);
      let _el$17 = _$createElement("div");
      _$setProp(_el$17, "className", "col-sm-6 smallpad");
      let _el$18 = _$createElement("button");
      _$setProp(_el$18, "type", "button");
      _$setProp(_el$18, "className", "btn btn-primary btn-block");
      _$setProp(_el$18, "id", "swaprows");
      _$setProp(_el$18, "onClick", () => this.swapRows());
      _$insertNode(_el$18, _$createTextNode("Swap Rows"));
      _$insertNode(_el$17, _el$18);
      _$insertNode(_el$6, _el$17);
      _$insertNode(_el$5, _el$6);
      _$insertNode(_el$2, _el$5);
      _$insertNode(_el$1, _el$2);
      _$insertNode(_el$, _el$1);
      let _el$19 = _$createElement("table");
      _$setProp(
        _el$19,
        "className",
        "table table-hover table-striped test-data",
      );
      let _el$20 = _$createElement("tbody");
      let _el$21 = _$createComponent(For, {
        each: data,
        fallback: _$createJsxTagElement(() => {
          let _el$23 = _$createElement("tr");
          return _el$23;
        }),
        children: [
          (item) =>
            _$createJsxTagElement(() => {
              let _el$24 = _$createComponent(Row, {
                item: item,
                selected: selected,
                onSelect: (id) => this.select(id),
                onRemove: (id) => this.remove(id),
              });
              return _el$24;
            }),
        ],
      });
      _$insertNode(_el$20, _el$21);
      _$insertNode(_el$19, _el$20);
      _$insertNode(_el$, _el$19);
      let _el$22 = _$createElement("span");
      _$setProp(_el$22, "className", "preloadicon glyphicon glyphicon-remove");
      _$setProp(_el$22, "aria-hidden", "true");
      _$insertNode(_el$, _el$22);
      return _el$;
    });
  }
}

export default App;

```
